server/src/main/java/nl/inl/blacklab/server/search/BlsCache.java

```diff
@@ -149,90 +172,104 @@ public void removeSearchesForIndex(BlackLabIndex index) {
-            if (future == null) {
-                search.log(LogLevel.BASIC, "not found in cache, starting search: " + search);
-                try {
-                    long freeMegs = MemoryUtil.getFree() / ONE_MB_BYTES;
-                    if (freeMegs < config.getMinFreeMemForSearchMegs()) {
-                        cleanupSearches();  // try to free up space for next search
-                        logger.warn(
-                                "Can't start new search, not enough memory (" + freeMegs + "M < "
-                                        + config.getMinFreeMemForSearchMegs() + "M)");
-                        logger.warn("(NOTE: make sure Tomcat's max heap mem is set to an appropriate value!)");
-                        throw new InsufficientMemoryAvailable(
-                                "The server has insufficient memory available to start a new search. Please try again later. (not enough JVM heap memory for new search; try increasing -Xmx value when starting JVM)");
-                    }
-                    // logger.debug("Enough free memory: " + freeMegs + "M"); // check that we have sufficient available memory
-                } catch (InsufficientMemoryAvailable e) {
-                    search.log(LogLevel.BASIC, "not enough memory for search: " + search + " (" + e.getMessage() + ")");
-                    throw e;
-                }
-                future = new BlsCacheEntry<>(search);
-                if (useCache)
-                    searches.put(search, future);
-                if (trace) logger.info("-- STARTING: {}", search);
+        future = useCache ? (BlsCacheEntry<R>) searches.get(search) : null;
+        if (future == null) {
+            search.log(LogLevel.BASIC, "not found in cache, starting search: " + search);
+            int numQueued = numberOfQueuedSearches();
+            if (numQueued >= config.getMaxQueuedSearches()) {
+                logger.warn("Can't start new search, too many queued searches (maxQueuedSearches = " + config.getMaxQueuedSearches() + ")");
+                throw new ServerOverloaded("The server is too busy right now (too many queued searches). Please try again later.");
+            }
+            future = new BlsCacheEntry<>(search);
+            if (useCache)
+                searches.put(search, future);
+
+            if (block) {
+                // Blocking search. Run it now.
+                traceInfo("-- STARTING: {} (BLOCKING SEARCH)", search);
                 future.start(block);
+            } else if (!allowQueue || !useCache) {
+                // No queueing allowed. Start the search right away.
+                // (we also do this if you bypass the cache, because then queueing doesn't work)
+                if (!allowQueue)
+                    traceInfo("-- STARTING: {} (TOP-LEVEL SEARCH)", search);
+                else
+                    traceInfo("-- STARTING: {} (NOT USING CACHE)", search);
+                future.startIfQueued();
```

Comment:
How can the future be queued if allowQueue is false?

Reply:
This is just to make the trace info clearer. If allowQueue is false, this is a top-level search (not an internal search). If it's true, then `usecache=no` was specified.

Reply:
Consider `allowQueue=false` and `useCache=false`, then `startIfQueued()` should be a is no-op, right? But then how does the future ever run? It's not in the queue, and it's not in the cache? Or is it up to the caller to start?

Reply:
Okay, I see. The missing bit of information is that _every `BlsCacheEntry` starts queued_ (i.e. not having been started). Then, if `!allowQueue`, we immediately start it by calling `startIfQueued()`. Essentially `isQueued()` could be written as `!searchWasStarted()`; maybe that should be the method instead. I'll consider that and at the very least improve the comments.

Reply:
(made quite a few changes to hopefully make everything clearer. See latest commits added to this pull request)